crates/exec-wasmtime/src/loader/attested.rs

```diff
@@ -1,25 +1,116 @@
+const TOML_MEDIA_TYPE: &str = "application/toml";
+const WASM_MEDIA_TYPE: &str = "application/wasm";
+
+fn get(url: impl AsRef<str>) -> Result<Response> {
+    let url = url.as_ref();
+    ureq::get(url)
+        .call()
+        .with_context(|| format!("failed to GET `{}`", url))
+}
+
+fn response_into_wasm(res: Response) -> Result<Vec<u8>> {
+    // TODO: Initialize with capacity of Content-Length if set.
+    let mut wasm = Vec::new();
+    res.into_reader()
+        .take(MAX_WASM_SIZE)
+        .read_to_end(&mut wasm)
+        .context("failed to read WASM module contents")?;
+    // TODO: Verify Content-Digest
+    Ok(wasm)
+}
+
+fn get_drawbridge_directory(url: impl AsRef<str>) -> Result<(Vec<u8>, String)> {
+    let url = url.as_ref().trim_end_matches('/');
+    // TODO: refactor into `get_typed<WASM_MEDIA_TYPE>` once `&str` is supported as
+    // const generic argument
+    let wasm = get(format!("{}/{}", url, "main.wasm"))
+        .and_then(|res| {
+            let typ = res.content_type();
+            ensure!(
+                typ == WASM_MEDIA_TYPE,
+                format!(
+                    "expected `main.wasm` to have `{}` media type, got `{}`",
+                    WASM_MEDIA_TYPE, typ
+                )
+            );
+            Ok(res)
+        })
+        .and_then(response_into_wasm)?;
+
+    // TODO: refactor into `get_typed<TOML_MEDIA_TYPE>` once `&str` is supported as
+    // const generic argument
+    let conf = get(format!("{}/{}", url, "Enarx.toml"))
+        .and_then(|res| {
+            let typ = res.content_type();
+            ensure!(
+                typ == TOML_MEDIA_TYPE,
+                format!(
+                    "expected `Enarx.toml` to have `{}` media type, got `{}`",
+                    TOML_MEDIA_TYPE, typ
+                )
+            );
+            Ok(res)
+        })?
+        // TODO: Verify Content-Digest
+        .into_string()?;
+
+    Ok((wasm, conf))
+}
 
 impl Loader<Attested> {
     pub fn next(self) -> Result<Loader<Acquired>> {
-        // TODO: get workload from drawbridge.
-
-        // FIXME: temporary hack to read wasm from fd 3.
-        let mut file = unsafe { std::fs::File::from_raw_fd(3) };
-        let mut webasm = Vec::new();
-        file.read_to_end(&mut webasm)?;
+        let (webasm, config) = match self.0.package {
+            Package::Remote(ref url) => {
+                let res = get(url.as_str())?;
+                match res.content_type() {
+                    WASM_MEDIA_TYPE => response_into_wasm(res).map(|webasm| (webasm, None))?,
+                    DRAWBRIDGE_DIRECTORY_MEDIA_TYPE => get_drawbridge_directory(url.as_str())
+                        .map(|(webasm, config)| (webasm, Some(config)))?,
+                    t => bail!("unsupported content type: {}", t),
+                }
+            }
+            Package::Local { wasm, conf } => {
+                let mut webasm = Vec::new();
+                unsafe { std::fs::File::from_raw_fd(wasm) }
```

Comment:
When we talk about safety here, what do we mean? Is it:
1. the fact that we're not going to segfault here that we're trying to assert
or,
2. are we talking about integrity of data and the whether or not we trust that this FD is indeed the FD containing WASM that the user wanted us to execute, i.e. no one has tampered with the data and this is indeed what the user wanted to execute.

Well, there's no way to programmatically validate that this is a valid FD AFAIK (@haraldh any ideas ?).
We assume that this executable was executed by the host binary, we assume that our implementation is correct, and, therefore, the FD opened by the host binary (via a safe operation) was passed to this executable and this operation *should* be safe. However, due to (2.), there's no guarantee that the opened FD by the host was not tampered with on the way.

We got this FD from the host and by definition we should not be trusting what host tells us, but we are trying to run a workload from a local FS here, which is insecure already and requires trust in the host anyway.
So basically, the only way this can be safe is if we trust the host and we don't, but I also don't think that this should really warrant marking `next()` unsafe, since that'd mean that essentially everything `exec-wasmtime` does is unsafe.

I'll keep the unsafe annotations as-is and add a SAFETY comment here as outlined above, but again, technically this whole thing is unsafe.